navigation/insilico_rotater.h

```diff
@@ -16,76 +24,140 @@ namespace Navigation {
-            rollImage(m_Image, m_BeginRoll);
-            if (!m_MaskImage.empty()) {
-                rollImage(m_MaskImage, m_BeginRoll);
+            const auto index = toIndex(beginRoll);
+            rollImage(image, m_Image, index);
+            if (!maskImage.empty()) {
+                rollImage(maskImage, m_MaskImage, index);
             }
         }
-        
-        size_t i = m_BeginRoll;
-        while (true) {
-            func(m_Image, m_MaskImage, i - m_BeginRoll);
-
-            i += m_ScanStep;
-            if (i >= m_EndRoll) {
-                break;
+
+        template<class Func>
+        void rotate(Func func)
+        {
+            auto i = m_BeginRoll;
+            while (true) {
+                func(m_Image, m_MaskImage, distance(m_BeginRoll, i));
+
+                i += m_ScanStep;
+                if (i >= m_EndRoll) {
+                    break;
+                }
+
+                const auto index = toIndex(i);
+                rollImage(m_ImageOriginal, m_Image, index);
+                if (!m_MaskImageOriginal.empty()) {
+                    rollImage(m_MaskImageOriginal, m_MaskImage, index);
+                }
             }
+        }
+
+        units::angle::radian_t columnToHeading(size_t column) const
+        {
+            return units::angle::turn_t{ (double) toIndex(column) / (double) m_ImageOriginal.cols };
+        }
+
+        size_t numRotations() const
+        {
+            return (m_EndRoll - m_BeginRoll) / m_ScanStep;
+        }
 
-            rollImage(m_Image, m_ScanStep);
-            if (!m_MaskImage.empty()) {
-                rollImage(m_MaskImage, m_ScanStep);
+    private:
+        const size_t m_ScanStep;
+        const IterType m_BeginRoll, m_EndRoll;
+        const cv::Mat &m_ImageOriginal, &m_MaskImageOriginal;
+        cv::Mat m_Image, m_MaskImage;
+
+        static void rollImage(const cv::Mat &imageIn, cv::Mat &imageOut, size_t pixels)
+        {
+            // Loop through rows
+            for (int y = 0; y < imageIn.rows; y++) {
+                // Get pointer to start of row
+                const uint8_t *rowPtr = imageIn.ptr(y);
+                uint8_t *rowPtrOut = imageOut.ptr(y);
+
+                // Rotate row to left by pixels
+                std::rotate_copy(rowPtr, rowPtr + pixels, rowPtr + imageIn.cols, rowPtrOut);
             }
         }
-    }
 
-    size_t max() const
+        static size_t distance(size_t first, size_t last)
+        {
+            return last - first;
+        }
+
+        template<typename Iter>
+        static size_t distance(Iter first, Iter last)
+        {
+            return static_cast<size_t>(std::distance(first, last));
+        }
+
+        static size_t toIndex(size_t index)
+        {
+            return index;
+        }
+
+        template<typename Iter>
+        static size_t toIndex(Iter it)
+        {
+            return *it;
+        }
+    };
+
+    template<typename IterType>
```

Comment:
Also I can't quite figure out what's gained by having these over the previous constructors with argument forwarding

Reply:
It's because it's now a template class, taking an ``IterType``, meaning if we didn't have the outer struct you'd have to use it like this:
```
PerfectMemoryRotater<..., ..., InSilicoRotater<some long iterator type>> pm;
```

By having it inside a struct and with the ``create()`` factory function the iterator type can just be inferred when e.g. ``getHeading()`` is called. Ah the fun of templates!

Reply:
BLEURGH!